contributors/design-proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,302 @@
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the value of the metric across the described objects
+    Items []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type MetricValue struct {
+    // a reference to the described object
+    DescribedObject ObjectReference `json:"describedObject"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // indicates the time at which the metrics were produced
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the window ([Timestamp-Window, Timestamp]) from
+    // which these metrics were calculated, when returning rate
+    // metrics calculated from cumulative metrics (or zero for
+    // non-calculated instantaneous metrics).
+    Window  unversioned.Duration `json:"window"`
+
+    // the value of the metric for this
+    Value resource.Quantity
+}
+```
+
+For instance, the example request above would yield the following object:
+
+```json
+{
+    "kind": "MetricValueList",
+    "apiVersion": "custom-metrics/v1alpha1",
+    "items": [
+        {
+            "metricName": "hits-per-second",
+            "describedObject": {
+                "kind": "Ingress",
+                "apiVersion": "extensions",
+                "name": "server1",
+                "namespace": "webapp"
+            },
+            "timestamp": SOME_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": "10"
+        },
+        {
+            "metricName": "hits-per-second",
+            "describedObject": {
+                "kind": "Ingress",
+                "apiVersion": "extensions",
+                "name": "server2",
+                "namespace": "webapp"
+            },
+            "timestamp": ANOTHER_TIMESTAMP_HERE,
+            "window": "10s",
+            "value": "15"
+        }
+    ]
+}
+```
+
+Semantics
+---------
+
+### Object Types ###
+
+In order to properly identify resources, we must use resource names
+qualified with group names (since the group for the requests will always
+be `custom-metrics`).
+
+The `object-type` parameter should be the string form of
+`unversioned.GroupResource`.  Note that we do not include version in this;
+we simply wish to uniquely identify all the different types of objects in
+Kubernetes.
+
+In the case of cross-group object renames, the adapter should maintain
+a list of "equivalent versions" that the monitoring system uses. This is
+monitoring-system dependent (for instance, the monitoring system might
+record all HorizontalPodAutoscalers as in `autoscaling`, but should be
+aware that HorizontalPodAutoscaler also exist in `extensions`).
+
+Note that for namespace metrics, we use a pseudo-resource called
+`metrics`.  Since there is no resource in the legacy API group, this will
+not clash with any existing resources.
+
+### Metric Names ###
+
+Metric names must be escaped so that any given metric appears as a single
```

Comment:
rather than try to allow a way to include an escaped "/" in an object name, kubernetes objects are simply not allowed to have names containing `/` or `%` characters, and are not allowed to be `.` or `..`

that allows you to use normal URL path segment encoding to safely plumb values through